events/diskpersist.go

```diff
@@ -265,71 +270,170 @@ const (
-	if len(p.evtbuf) == 0 {
+func (dp *DiskPersistence) flushLog(ctx context.Context) error {
+	if len(dp.evtbuf) == 0 {
 		return nil
 	}
 
-	_, err := io.Copy(p.logfi, p.outbuf)
+	_, err := io.Copy(dp.logfi, dp.outbuf)
 	if err != nil {
 		return err
 	}
 
-	p.outbuf.Truncate(0)
+	dp.outbuf.Truncate(0)
 
-	for _, ej := range p.evtbuf {
-		p.broadcast(ej.Evt)
+	for _, ej := range dp.evtbuf {
+		dp.broadcast(ej.Evt)
 		ej.Buffer.Truncate(0)
-		p.buffers.Put(ej.Buffer)
+		dp.buffers.Put(ej.Buffer)
 	}
 
-	p.evtbuf = p.evtbuf[:0]
+	dp.evtbuf = dp.evtbuf[:0]
 
 	return nil
 }
 
-func (p *DiskPersistence) doPersist(j persistJob) error {
-	b := j.Buf
+func (dp *DiskPersistence) garbageCollectRoutine() {
+	t := time.NewTicker(time.Hour)
+
+	for {
+		ctx := context.Background()
+		select {
+		// Closing a channel can be listened to with multiple routines: https://goplay.tools/snippet/UcwbC0CeJAL
+		case <-dp.shutdown:
+			return
+		case <-t.C:
+			if errs := dp.garbageCollect(ctx); len(errs) > 0 {
+				for _, err := range errs {
+					log.Errorf("garbage collection error: %s", err)
+				}
+			}
+		}
+	}
+}
+
+var garbageCollectionsExecuted = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "disk_persister_garbage_collections_executed",
+	Help: "Number of garbage collections executed",
+}, []string{})
+
+var garbageCollectionErrors = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "disk_persister_garbage_collections_errors",
+	Help: "Number of errors encountered during garbage collection",
+}, []string{})
+
+var refsGarbageCollected = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "disk_persister_garbage_collections_refs_collected",
+	Help: "Number of refs collected during garbage collection",
+}, []string{})
+
+var filesGarbageCollected = promauto.NewCounterVec(prometheus.CounterOpts{
+	Name: "disk_persister_garbage_collections_files_collected",
+	Help: "Number of files collected during garbage collection",
+}, []string{})
+
+func (dp *DiskPersistence) garbageCollect(ctx context.Context) []error {
+	garbageCollectionsExecuted.WithLabelValues().Inc()
+
+	// Grab refs created before the retention period
+	var refs []LogFileRef
+	var errs []error
+
+	defer func() {
+		garbageCollectionErrors.WithLabelValues().Add(float64(len(errs)))
+	}()
+
+	if err := dp.meta.WithContext(ctx).Find(&refs, "created_at < ?", time.Now().Add(-dp.retention)).Error; err != nil {
+		return []error{err}
+	}
+
+	oldRefsFound := len(refs)
+	refsDeleted := 0
+	filesDeleted := 0
+
+	// In the future if we want to support Archiving, we could do that here instead of deleting
+	for _, r := range refs {
+		dp.lk.Lock()
+		currentLogfile := dp.logfi.Name()
+		dp.lk.Unlock()
+
+		if filepath.Join(dp.primaryDir, r.Path) == currentLogfile {
+			// Don't delete the current log file
+			log.Info("skipping deletion of current log file")
+			continue
+		}
+
+		// Delete the ref in the database to prevent playback from finding it
```

Comment:
I would swap the order of deleting the file and the ref. If we fail to delete the file and havent deleted the ref, then we can just try again later once we've rectified whatever issue was there.

though I guess if someone requested the events in the file after we deleted it but before we dropped the ref that would be weird for them.

Reply:
I want to delete the ref first because if we delete the file but fail to delete the ref, we could try to open it for playback and fail.

Reply:
We can clean up old files with a cron job if we need to